### PR TITLE
Expose a way to override kSecAttrService

### PR DESCRIPTION
**Pull request by Autoc0diq**
Legacy app may have created shared keychain using a different service than the hardcoded one. This allows to specify it.

### CODE REVIEW

**Comment by dfed** [comment]
Apps migrating to Valet should use `migrateObjectsMatchingQuery:removeOnCompletion:` to bring existing keychain items into Valet, rather than attempting to make Valet work with the existing identifier/service combo.

In case you're curious here's why Valet sets the service internally:
• It ensures that changing Accessibility and storage type (iCloud vs local, etc) metadata works 100% of the time. Updating metadata on existing keychain entities has a ~0.5% failure rate, and the failures are silent. Valet avoids this issue by relying on copying the data from one service to another (which does not fail), rather than updating in place. In fact, our migration code would fail if Valet if the service were set manually, as updating metadata would result in `errSecDuplicateItem` since Valet is not built to update items in place.
• It ensures that each Valet lives in a sandbox – Valets that do not pass the isEqual: test should not be able to stomp one another's values.
• It keeps external adopters of Valet from needing to grapple with multiple unique identifiers (service vs identifier, etc).
• It gives Valet enough control over its ecosystem to ensure that operations will fail only when you access to the keychain is denied.

Hopefully that elucidates why Valet does what it does :). Do let us know if you run into issues using `migrateObjectsMatchingQuery:removeOnCompletion:`.

**Comment by Autoc0diq** [comment]
Thanks for the info. Let me explain why we need this.
We have a legacy app that exposes some data via a shared access group. Unfortunately this app is now maintained and cannot be updated to use Valet in the field. This app was using a certain kAttrService. Without a way to specify the kAttrService when creating a VALValet, there would be no way to even just read that data.
The solution to migrate the data is possible but not ideal. We would have to migrate the data every time we try to access that shared access group.
I think exposing that method in a class extension would be a good compromise.

**Comment by Autoc0diq** [comment]
Side note. None of the tests exercise the initializer with the sharedAccessGroup. Some tests are actually not doing what the test method name is saying i.e.: test_stringForKey_differentAccessGroupFailsToRetrieveString

It also appears that the method _sharedAccessGroupPrefix is returning an empty bundleSeedID when running the tests after trying to add a test for sharedAccessGroup.

**Comment by dfed** [comment]
You can migrate the data in using `migrateObjectsMatchingQuery:removeOnCompletion:`, no? e.g. `[valet migrationObjectsMatchingQuery:{(__bridge id)kAttrService : @"oldServiceName"} removeOnCompletion:YES]`. Valet will then bring that old keychain value into your new Valet, and then there's no need to modify Valet.

The reason none of the tests test the `sharedAccessGroup` initializer is that Xcode's test environment doesn't really work with shared access groups. `_sharedAccessGroupPrefix` returning `nil` in the Xcode 7 test environment is a known issue. See #33.

As for `test_stringForKey_differentAccessGroupFailsToRetrieveString`... good find! I'll look into that. Any others you found?

**Comment by Autoc0diq** [comment]
As I explained, our legacy app still needs the old data in the current place with given kAttrService as such  we cannot migrate the data. We could and leave the data in place by passing NO to removeOnCompletion but we would have to run this migration everytime we run our new app in case the shared data was modified by our legacy app. Seems like a hack when a simple new initializer could be exposed.

**Comment by dfed** [comment]
Ah I didn't realize that you needed to keep the old and new keychain data in sync rather than doing a direct forward migration.

The new initializer violates a central tenant to the Valet architecture (see the explanation above, though I'm happy to elaborate on any of the points), and would give adoptees of Valet rope to hang themselves with. For example, if you were to have a Valet that read the old keychain data and then you attempted to migrate that Valet to a new Valet with the same service but different metadata, your migration would fail.

Migrating the old data each app launch is your best bet.

**Comment by Autoc0diq** [comment]
What type of metadata? Accessibility? I think it would be fine to fail the migration in this case. Is that the only use case?

Wouldn't the same situation already occur if you were to use the existing method migrateObjectsMatchingQuery:removeOnCompletion passing a kAttrService set to the new valet internal service name and different accessibility values?

**Comment by dfed** [comment]
The "metadata" refers to accessibility, Valet type (secure enclave vs iCloud storage vs regular storage), and whether the keychain items are stored with a shared access group or not. Again, the fact that the service name is derived internally is a key component in how Valet is able to guarantee that Valet will only fail to retrieve data when the keychain is inaccessible. It is essential to our sandboxing and threading model, in addition to our migration model.

Moreover, setting the service name manually renders the `identifier` ivar meaningless for Valets that are not created with a shared access group.

`migrateObjectsMatchingQuery:removeOnCompletion:` isn't meant to be used when migration between Valet objects (at least, not externally), that's why we wrote `migrationObjectsFromValet:removeOnCompletion:`. If you're passing in a `kAttrService` that'll match the new valet internal service name (which is prefixed with `VAL_` so as not to conflict with service names that are used by developers in the wild), it's safe to say you're doing it wrong.

In short, the use case you have – interoperability with an app that can not be updated to use Valet – is not a supported case for Valet, nor is it a desired case for Valet. If you'd like Valet to support this use case in your app, I recommend subclassing VALValet and using your subclass, or using a Category on VALValet to let you set the service name and update the baseQuery.
